sdks/python/apache_beam/io/filebasedsink.py

```diff
@@ -178,85 +178,116 @@ def open_writer(self, init_result, uid):
-          if exception:
-            logging.warning('Rename not successful: %s -> %s, %s', src, dest,
-                            exception)
-            should_report = True
-            if isinstance(exception, IOError):
-              # May have already been copied.
-              try:
-                if FileSystems.exists(dest):
-                  should_report = False
-              except Exception as exists_e:  # pylint: disable=broad-except
-                logging.warning('Exception when checking if file %s exists: '
-                                '%s', dest, exists_e)
-            if should_report:
-              logging.warning(('Exception in _rename_batch. src: %s, '
-                               'dest: %s, err: %s'), src, dest, exception)
-              exceptions.append(exception)
-          else:
-            logging.debug('Rename successful: %s -> %s', src, dest)
-        return exceptions
-
-    exception_batches = util.run_using_threadpool(
-        _rename_batch, zip(source_file_batch, destination_file_batch),
-        num_threads)
+    existing_files = []
+    for shard_num in xrange(len(writer_results)):
+      final_name = self._get_final_name(shard_num, num_shards)
+      if FileSystems.exists(final_name):
+        existing_files.append(final_name)
+    if existing_files:
+      logging.info('Deleting existing files in target path: %d',
+                   len(existing_files))
+      FileSystems.delete(existing_files)
 
-    all_exceptions = [e for exception_batch in exception_batches
-                      for e in exception_batch]
-    if all_exceptions:
-      raise Exception('Encountered exceptions in finalize_write: %s',
-                      all_exceptions)
+  @check_accessible(['file_path_prefix'])
+  def finalize_write(self, init_result, writer_results,
+                     unused_pre_finalize_results):
+    writer_results = sorted(writer_results)
+    num_shards = len(writer_results)
 
-    for final_name in destination_files:
-      yield final_name
+    src_files = []
+    dst_files = []
+    chunk_size = FileSystems.get_chunk_size(self.file_path_prefix.get())
+    num_skipped = 0
+    for shard_num, shard in enumerate(writer_results):
+      final_name = self._get_final_name(shard_num, num_shards)
+      src = shard
+      dst = final_name
+      src_exists = FileSystems.exists(src)
+      dst_exists = FileSystems.exists(dst)
+      if not src_exists and not dst_exists:
+        raise BeamIOError('src and dst files do not exist. src: %s, dst: %s' % (
+            src, dst))
+      if not src_exists and dst_exists:
+        logging.debug('src: %s -> dst: %s already renamed, skipping', src, dst)
+        num_skipped += 1
+        continue
+      if (src_exists and dst_exists and
+          FileSystems.checksum(src) == FileSystems.checksum(dst)):
+        logging.debug('src: %s == dst: %s, deleting src', src, dst)
+        FileSystems.delete([src])
```

Comment:
<!--new_thread; commit:bc394b7dd981af5640e2e80ebcc83e86eee0136e; resolved:0-->
Can we do a bulk call ?

Reply:
Done